File: bs_comparison/bs.py
import numpy as np
# import pandas as pd
# import pandas_datareader as pdr
# from scipy import stats
import matplotlib.pyplot as plt
# import seaborn as sns

np.random.seed(20)

def bsm_call(S_t, K, r, vol, T, t=0):
	den = 1 / (vol * np.sqrt(T - t))
	d1 = den * (np.log(S_t / K) + (r + 0.5 * vol ** 2) * (T - t))
	d2 = den * (np.log(S_t / K) + (r - 0.5 * vol ** 2) * (T - t))
	C = S_t * stats.norm.cdf(d1) \
	- K * np.exp(-r * (T - t)) * stats.norm.cdf(d2)
	return C

def bsm_call_mc(S_0, K, r, vol, T):
	# Number of time steps for simulation
	n_steps = int(T * 252)
	# Time interval
	dt = T / n_steps
	# Number of simulations
	N = 100000
	# Zero array to store values (often faster than appending)
	S = np.zeros((n_steps, N))
	S[0] = S_0

	for t in range(n_steps-1):
		# Draw random values to simulate Brownian motion
		Z = np.random.standard_normal(N)
		S[t+1] = S[t] * np.exp((r - 0.5 * vol ** 2) * dt + (vol * np.sqrt(dt) * Z))

	# Sum and discount values
	C = np.exp(-r * T) * 1 / N * np.sum(np.maximum(S[-1] - K, 0))
	# plt.figure(figsize=(12,8))
	# plt.plot(S[:,0:50])
	# plt.axhline(K, c="k", xmin=0,
	# 			xmax=n_steps,
	# 		label="Strike Price")
	# plt.xlim([0, n_steps])
	# plt.ylabel("Non-Discounted Value")
	# plt.xlabel("Time step")
	# plt.title("First 50 Option Paths")
	# plt.legend(loc="best")
	# plt.show()
	return C


# Parameters - same values as used in the example above 
# repeated here for a reminder, change as you like
# Initial asset price
S_0 = 100
# Strike price for call option
K = 110
# Time to maturity in years
T = 0.5
# Risk-free rate of interest
r = 0.03
# Historical Volatility
vol = 0.5

# C_bsm = bsm_call(S_0, K, r, vol, T)
C = bsm_call_mc(S_0, K, r, vol, T)

print("Strike price: {:.2f}".format(K))
# print("BSM Option Value Estimate: {:.2f}".format(C_bsm))
print("Monte Carlo Option Value Estimate: {:.2f}".format(C))

File: bs_comparison/bs.js
// returns a gaussian random function with the given mean and stdev.
function gaussian(mean, stdev) {
    var y2;
    var use_last = false;
    return function() {
        var y1;
        if(use_last) {
           y1 = y2;
           use_last = false;
        }
        else {
            var x1, x2, w;
            do {
                 x1 = 2.0 * Math.random() - 1.0;
                 x2 = 2.0 * Math.random() - 1.0;
                 w  = x1 * x1 + x2 * x2;               
            } while( w >= 1.0 || w == 0);
            w = Math.sqrt(-2.0 * Math.log(w) / w);
            y1 = x1 * w;
            y2 = x2 * w;
            use_last = true;
       }

       var retval = mean + stdev * y1;
       return retval;
   }
}

function BlackScholes(PutCallFlag, S, X, T, r, v) {

    var d1, d2;
    d1 = (Math.log(S / X) + (r + v * v / 2.0) * T) / (v * Math.sqrt(T));
    d2 = d1 - v * Math.sqrt(T);


    if (PutCallFlag== "c")
        return S * CND(d1)-X * Math.exp(-r * T) * CND(d2);
    else
        return X * Math.exp(-r * T) * CND(-d2) - S * CND(-d1);

}

/* The cummulative Normal distribution function: */

function CND(x){

    var a1, a2, a3, a4 ,a5, k ;

    a1 = 0.31938153, a2 =-0.356563782, a3 = 1.781477937, a4= -1.821255978 , a5= 1.330274429;

    if(x<0.0)
        return 1-CND(-x);
    else
        k = 1.0 / (1.0 + 0.2316419 * x);
    return 1.0 - Math.exp(-x * x / 2.0)/ Math.sqrt(2*Math.PI) * k
        * (a1 + k * (-0.356563782 + k * (1.781477937 + k * (-1.821255978 + k * 1.330274429)))) ;

}

function BsmCallMc(S_t, K, r, vol, T) {
    // Number of time steps for simulation
	var n_steps = Math.floor(T * 252)
	// Time interval
	var dt = T / n_steps
	// Number of simulations
    var N = 100000
    // Zero array to store values (often faster than appending)
    var S = Array(n_steps).fill().map(() => Array(N).fill(S_t))
    var standard = gaussian(0, 1);

    for (var t=0; t<n_steps-1; t++) {
        for (var u=0; u<N; u++) {
            S[t+1][u] = S[t][u] * Math.exp((r - 0.5 * Math.pow(vol, 2)) * dt + (vol * Math.sqrt(dt) * standard()))
        }
    }
    
    S.map(x=>
        console.log(x[N-1])
    )
    var C = Math.exp(-r * T) * 1 / N * Math.max( S.map(x=>x[N-1]-K), 0 )
    return C
}

var S_0 = 100
var K = 110
var T = 0.5
var r = 0.03
var vol = 0.5

var C = BsmCallMc(S_0, K, r, vol, T)

console.log("Strike price: " + K.toFixed(2))
console.log("Monte Carlo Option Value Estimate: " + C.toFixed(2))